Improve calculation of average and more fail safety

kscholty opened the pull request:
I propose multiple changes here

1) Set the switch to active low and pull up the input pin. This will set the drive to neutral if the switch gets disconnected or fails.
    I feel this is the safer option than having the pin floating without the switch. Of course an external pull-down resistor could achive that  with the original code as well.
2) In order to switch to NEUTRAL immediately if the ermergency switch is flipped, modify the average values so that the switch happens without smoothing.
2) Improve the calculation of the average of the last values. It's not necessary to calculate the sum over and over again. 
    This simplifies the code and speeds up the calculation. Furthermore, since it's only about smoothing, i don't think the we have to bother the poor chip with floating point calculations if not necessary.
3) Initializte the DAC with the neutral value in setup. Not really important but looks like a safe choice to me.

I created this project on WOKWI for you. This allows you to test the code without the need for real hardware. Quite convenient in early stages of development.

Its called [DigimerHelmControl](https://wokwi.com/projects/388912131648950273)

kscholty committed:
Improve calculation of average
Set switch to active low to improve fault behavior

digimer commented:
Thank you very much for this! I'll review and test shortly. :D

kscholty committed:
Smoother transition from Switch off to On
and even simpler calculation of averages, due to starting with a full array

dejanv commented:
Hi Maddie
I suggest that you fill up the potArray in void setup() function as this is done only once. Except if you want to remove all values in some emergency situations.
KCSHOLTY already mention better sum calculation.
in setup() also calculate the potTotal and then in loop() just remove old and add new value. No need to summarize the potTotal over and over again. 
potTotal = potTotal - potArray[oldestPotIndex] + sensorValue;
potArray[oldestPotIndex] = sensorValue;
  oldestPotIndex = oldestPotIndex + 1;
  if (oldestPotIndex >= averageOver) {
 ... ETC ...

The HW is very unpredictable this valid also for digitalRead() function that reads variable resistor.
It could happened that your engine will be feed with two flipped voltage values.  Question is what will engine do.
Let wait for first engine test and then react appropriately to avoid over design in advance.
Idea is to do a smoothly  multi steps change from old value to new value.

kscholty commented:
Initializing the array is what I committed a few hours ago. Makes the code even simpler.
This can also be used to handle the switch better. 
Using an interrupt for the switch would be even better, but I didn't want to make it too complicated.

dejanv commented on helm-control:
```diff
@@ -66,64 +73,70 @@ static const int32_t reverseSensorSteps = (reverseMinimum - reverseMaximum);
   // Read the helm position potentiometer if the main switch is on
   int32_t sensorValue;
-  if (mainSwitchVal == 0) {
+  if (mainSwitchVal == 1) {
```
if (mainSwitchVal == 1) is ok, 
exist a shorter version as every value greater than zero is TRUE
if (mainSwitchVal)

kscholty replied:
Which is semantically the same but harder to understand for unexperienced programmers, which is why I put it that way.

digimer replied:
Thankfully for me, perl also uses the simpler if ($var) { ... } format, so I can read the shorter version. That said, future folks who might want to use this to start on their own projects might benefit from it. I assume it's no real expense, in running or compiled size?

kscholty replied:
The created machine code should be identical.

dejanv reviewed (commented):
:) I start commenting the main code and then notice a fresh pull request.

dejanv commented on helm-control:
```diff
@@ -66,64 +73,70 @@ static const int32_t reverseSensorSteps = (reverseMinimum - reverseMaximum);
-int32_t potLoopCount     = 0;
 int32_t potArray[averageOver];
 
+void initializeAverages() {
+  for (int i = 0; i < averageOver; ++i) {
+    potArray[i] = neutralMidPoint;
+  }
+  potTotal = averageOver * neutralMidPoint;
+  oldestPotIndex = 0;
+}
+
```
I like the initialization to neutral value, but as soon the Loop() function will started then handle values arrives and in case handle is in forward or reverse position then engine will produce trust immediately.
On the power on the program must send only neutral values to engine until handle is really in neutral position. This prove that owner handle the boat.

kscholty replied:
I tried to stick with optimizing the existing code, not coming up with completely new functionality.

dejanv reviewed (commented):
Initial trust value must be controlled to prevent unintentional powering the engine.

digimer commented:
Editing to add; It's not the variable, I manually set '```delay(100);```' with no change.

digimer commented:
Before you do any work, I've got a lot of changes coming in first-start (https://github.com/digimer/Mermaids-Rest/tree/first-start) branch. I really like your idea of the array initialization and I'll copy that over to that branch. Once I'm done with it (later this week, it's got changes needed for the motor first start), then it might be easier to see what's in your PR that I've not copied over yet. :)

kscholty committed:
DAC init was commented out

DAC initialization was deactivated, probably for tests with the simulator. This probably caused a crash of the code when setting the DAC value.

kscholty committed:
Set initial DAC value

Setting first output of the DAC in setup already. Required to move setup() below the constants and variables.

kscholty commented:
> Editing to add; It's not the variable, I manually set '`delay(100);`' with no change.

It's probably a crash caused by me commenting out the DAC initialization in setup() for my tests. I corrected that. Should work now.
Sorry for that stupid error.

digimer commented:
I got the pre-population logic into the most recent PR. I'll be reviewing your other improvements next and merging in anything else. Sorry I didn't merge this PR directly, I made a LOT of changes so the conflict resolution would be ... a lot. :) Just the same, your contributions are getting into main though. :D

digimer commented:
No longer relevant after the refactor of PR 20, so closing.
